File: apps/mobile/client-apple/VettaKit/Sources/VettaKit/Protocol/PairingURI.swift
import Foundation

/// Everything a phone needs to reach one desktop, carried by the QR code (port of `pairing-uri.ts`).
public struct RemotePairingInvite: Equatable, Sendable {
	public static let version = 2
	public var pairingId: String
	public var mobileSecret: String
	public var desktopIdentityKey: String
	public var desktopName: String
	/// `host:port` pairs reachable on the local network, most preferred first.
	public var lanEndpoints: [String]
	/// Relay base URL (`wss://host`), nil when the desktop keeps cloud access off.
	public var relayBaseUrl: String?

	public init(pairingId: String, mobileSecret: String, desktopIdentityKey: String, desktopName: String, lanEndpoints: [String], relayBaseUrl: String? = nil) {
		self.pairingId = pairingId
		self.mobileSecret = mobileSecret
		self.desktopIdentityKey = desktopIdentityKey
		self.desktopName = desktopName
		self.lanEndpoints = lanEndpoints
		self.relayBaseUrl = relayBaseUrl
	}
}

public enum PairingURI {
	public static let scheme = "vetta"
	public static let host = "pair"

	private static let idPattern = try! NSRegularExpression(pattern: "^[A-Za-z0-9_-]{16,128}$")
	private static let hostPortPattern = try! NSRegularExpression(pattern: "^(\\[[0-9a-fA-F:.%a-zA-Z]+\\]|[A-Za-z0-9.-]+):(\\d{1,5})$")

	public static func build(_ invite: RemotePairingInvite) -> String {
		var params: [(String, String)] = [
			("v", String(RemotePairingInvite.version)),
			("id", invite.pairingId),
			("s", invite.mobileSecret),
			("k", invite.desktopIdentityKey),
			("n", invite.desktopName),
		]
		if !invite.lanEndpoints.isEmpty { params.append(("lan", invite.lanEndpoints.joined(separator: ","))) }
		if let relay = invite.relayBaseUrl, !relay.isEmpty { params.append(("relay", relay)) }
		let query = params.map { "\(formEncode($0.0))=\(formEncode($0.1))" }.joined(separator: "&")
		return "\(scheme)://\(host)?\(query)"
	}

	public static func parse(_ text: String) throws -> RemotePairingInvite {
		let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
		guard let components = URLComponents(string: trimmed), components.scheme != nil else {
			throw RemoteProtocolError("pairing link is not a valid URL")
		}
		guard components.scheme?.lowercased() == scheme, components.host?.lowercased() == host || components.path == "//\(host)" else {
			throw RemoteProtocolError("pairing link must start with vetta://pair")
		}
		let params = formDecode(components.percentEncodedQuery ?? "")
		guard params["v"] == String(RemotePairingInvite.version) else {
			throw RemoteProtocolError("pairing link version is unsupported")
		}
		let pairingId = params["id"] ?? ""
		let mobileSecret = params["s"] ?? ""
		let desktopIdentityKey = params["k"] ?? ""
		let desktopName = (params["n"] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
		guard matches(idPattern, pairingId) else { throw RemoteProtocolError("pairing link id is invalid") }
		guard matches(idPattern, mobileSecret) else { throw RemoteProtocolError("pairing link secret is invalid") }
		_ = try RemoteCrypto.decodePublicKey(desktopIdentityKey, field: "desktop identity key")
		guard !desktopName.isEmpty, desktopName.utf16.count <= 128 else {
			throw RemoteProtocolError("pairing link device name is invalid")
		}
		let lanEndpoints = (params["lan"] ?? "")
			.split(separator: ",", omittingEmptySubsequences: false)
			.map { $0.trimmingCharacters(in: .whitespaces) }
			.filter { !$0.isEmpty }
		for endpoint in lanEndpoints where !isValidHostPort(endpoint) {
			throw RemoteProtocolError("pairing link LAN endpoint is invalid")
		}
		return RemotePairingInvite(
			pairingId: pairingId,
			mobileSecret: mobileSecret,
			desktopIdentityKey: desktopIdentityKey,
			desktopName: desktopName,
			lanEndpoints: lanEndpoints,
			relayBaseUrl: normalizeRelayBaseUrl(params["relay"])
		)
	}

	public static func isValidHostPort(_ value: String) -> Bool {
		let range = NSRange(value.startIndex..., in: value)
		guard let match = hostPortPattern.firstMatch(in: value, range: range),
		      let portRange = Range(match.range(at: 2), in: value),
		      let port = Int(value[portRange])
		else { return false }
		return port >= 1 && port <= 65_535
	}

	/// Accepts http(s)/ws(s) and returns a `ws(s)://host[/path]` base without a trailing slash.
	public static func normalizeRelayBaseUrl(_ value: String?) -> String? {
		guard let value, !value.isEmpty,
		      let components = URLComponents(string: value.trimmingCharacters(in: .whitespacesAndNewlines)),
		      let host = components.host, !host.isEmpty
		else { return nil }
		let scheme: String
		switch components.scheme?.lowercased() {
		case "http", "ws": scheme = "ws"
		case "https", "wss": scheme = "wss"
		default: return nil
		}
		let port = components.port.map { ":\($0)" } ?? ""
		let path = components.percentEncodedPath.isEmpty ? "/" : components.percentEncodedPath
		var base = "\(scheme)://\(host.lowercased())\(port)\(path)"
		if base.hasSuffix("/") { base.removeLast() }
		return base
	}

	public static func lanControlUrl(endpoint: String, pairingId: String) -> String {
		"ws://\(endpoint)/v2/lan/\(uriComponent(pairingId))"
	}

	public static func relayControlUrl(relayBaseUrl: String, pairingId: String, role: RemoteRole) -> String {
		"\(relayBaseUrl)/v2/relay/\(uriComponent(pairingId))/\(role.rawValue)"
	}

	static func uriComponent(_ text: String) -> String {
		var allowed = CharacterSet.alphanumerics
		allowed.insert(charactersIn: "-_.!~*'()")
		return text.addingPercentEncoding(withAllowedCharacters: allowed) ?? text
	}

	/// application/x-www-form-urlencoded, as URLSearchParams serializes.
	static func formEncode(_ text: String) -> String {
		var allowed = CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")
		allowed.insert(charactersIn: "*-._ ")
		let encoded = text.addingPercentEncoding(withAllowedCharacters: allowed) ?? text
		return encoded.replacingOccurrences(of: " ", with: "+")
	}

	/// URLSearchParams semantics: `+` is a space and the first value wins for `get`.
	static func formDecode(_ query: String) -> [String: String] {
		var result: [String: String] = [:]
		for pair in query.split(separator: "&", omittingEmptySubsequences: true) {
			let parts = pair.split(separator: "=", maxSplits: 1, omittingEmptySubsequences: false)
			let decode = { (raw: Substring) -> String in
				let spaced = raw.replacingOccurrences(of: "+", with: " ")
				return spaced.removingPercentEncoding ?? spaced
			}
			let key = decode(parts[0])
			let value = parts.count > 1 ? decode(parts[1]) : ""
			if result[key] == nil { result[key] = value }
		}
		return result
	}
}

File: packages/remote-control/src/pairing-uri.ts
import { decodePublicKey } from "./crypto.js";
import { RemoteProtocolError } from "./protocol.js";

export const PAIRING_URI_SCHEME = "vetta";
export const PAIRING_URI_HOST = "pair";
export const PAIRING_URI_VERSION = 2;

/**
 * Everything a phone needs to reach one desktop, carried by the QR code or
 * typed in by hand. The mobile secret is a long-lived per-device credential:
 * the desktop pins the first identity key that presents it and rejects any
 * other, so a photographed code cannot be reused once the real phone paired.
 */
export interface RemotePairingInvite {
	readonly version: typeof PAIRING_URI_VERSION;
	readonly pairingId: string;
	readonly mobileSecret: string;
	readonly desktopIdentityKey: string;
	readonly desktopName: string;
	/** `host:port` pairs reachable on the local network, most preferred first. */
	readonly lanEndpoints: readonly string[];
	/** Relay base URL (`wss://host`) or undefined when the desktop keeps cloud access off. */
	readonly relayBaseUrl?: string;
}

const ID_PATTERN = /^[A-Za-z0-9_-]{16,128}$/;
const HOST_PORT_PATTERN = /^(\[[0-9a-fA-F:.%a-zA-Z]+\]|[A-Za-z0-9.-]+):(\d{1,5})$/;

export function buildPairingUri(invite: RemotePairingInvite): string {
	const params = new URLSearchParams();
	params.set("v", String(invite.version));
	params.set("id", invite.pairingId);
	params.set("s", invite.mobileSecret);
	params.set("k", invite.desktopIdentityKey);
	params.set("n", invite.desktopName);
	if (invite.lanEndpoints.length > 0) params.set("lan", invite.lanEndpoints.join(","));
	if (invite.relayBaseUrl) params.set("relay", invite.relayBaseUrl);
	return `${PAIRING_URI_SCHEME}://${PAIRING_URI_HOST}?${params.toString()}`;
}

export function parsePairingUri(text: string): RemotePairingInvite {
	let url: URL;
	try {
		url = new URL(text.trim());
	} catch {
		throw new RemoteProtocolError("pairing link is not a valid URL");
	}
	if (
		url.protocol !== `${PAIRING_URI_SCHEME}:` ||
		(url.host !== PAIRING_URI_HOST && url.pathname !== `//${PAIRING_URI_HOST}`)
	) {
		throw new RemoteProtocolError("pairing link must start with vetta://pair");
	}
	const params = url.searchParams;
	if (params.get("v") !== String(PAIRING_URI_VERSION))
		throw new RemoteProtocolError("pairing link version is unsupported");
	const pairingId = params.get("id") ?? "";
	const mobileSecret = params.get("s") ?? "";
	const desktopIdentityKey = params.get("k") ?? "";
	const desktopName = (params.get("n") ?? "").trim();
	if (!ID_PATTERN.test(pairingId)) throw new RemoteProtocolError("pairing link id is invalid");
	if (!ID_PATTERN.test(mobileSecret)) throw new RemoteProtocolError("pairing link secret is invalid");
	decodePublicKey(desktopIdentityKey, "desktop identity key");
	if (!desktopName || desktopName.length > 128) throw new RemoteProtocolError("pairing link device name is invalid");
	const lanEndpoints = (params.get("lan") ?? "")
		.split(",")
		.map((entry) => entry.trim())
		.filter(Boolean);
	for (const endpoint of lanEndpoints) {
		if (!isValidHostPort(endpoint)) throw new RemoteProtocolError("pairing link LAN endpoint is invalid");
	}
	const relayBaseUrl = normalizeRelayBaseUrl(params.get("relay") ?? undefined);
	return {
		version: PAIRING_URI_VERSION,
		pairingId,
		mobileSecret,
		desktopIdentityKey,
		desktopName,
		lanEndpoints,
		...(relayBaseUrl ? { relayBaseUrl } : {}),
	};
}

export function isValidHostPort(value: string): boolean {
	const match = HOST_PORT_PATTERN.exec(value);
	if (!match) return false;
	const port = Number(match[2]);
	return port >= 1 && port <= 65_535;
}

/** Accepts http(s)/ws(s) and returns a `ws(s)://host[/path]` base without a trailing slash. */
export function normalizeRelayBaseUrl(value: string | undefined): string | undefined {
	if (!value) return undefined;
	let parsed: URL;
	try {
		parsed = new URL(value.trim());
	} catch {
		return undefined;
	}
	const protocol =
		parsed.protocol === "http:" || parsed.protocol === "ws:"
			? "ws:"
			: parsed.protocol === "https:" || parsed.protocol === "wss:"
				? "wss:"
				: undefined;
	if (!protocol) return undefined;
	return `${protocol}//${parsed.host}${parsed.pathname}`.replace(/\/$/, "");
}

export function lanControlUrl(endpoint: string, pairingId: string): string {
	return `ws://${endpoint}/v2/lan/${encodeURIComponent(pairingId)}`;
}

export function relayControlUrl(relayBaseUrl: string, pairingId: string, role: "mobile" | "desktop"): string {
	return `${relayBaseUrl}/v2/relay/${encodeURIComponent(pairingId)}/${role}`;
}
